In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder

from sklearn.neural_network import MLPClassifier, MLPRegressor

from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [2]:
# Load in the dataset

df =  pd.read_csv('http://bit.ly/diabetes_data')
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [4]:
# Scale the data

# Split into features and labels

X = df.loc[:, 'Pregnancies' : 'Age'].values
y = df['Outcome'].values

# Split further into training and testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Scale the training and testing values

# Scale the values
scaler = StandardScaler()

#Fit the model
scaler.fit(X_train)

# Transform the values
X_train = scaler.transform(X_train)

X_test = scaler.transform(X_test)


In [22]:
# Call the Neural Networks model and fit data

NN = MLPClassifier(hidden_layer_sizes=(15, 15, 15), activation='tanh', max_iter = 700)

NN.fit(X_train, y_train)

MLPClassifier(activation='tanh', hidden_layer_sizes=(15, 15, 15), max_iter=700)

In [23]:
# Predictions and Preformance evaluation

# Predictions 

y_pred = NN.predict(X_test)

# Performance

print('Accuracy score : ', accuracy_score(y_test, y_pred))

print('-----------------------------------------------')

print (confusion_matrix(y_test,y_pred))

print('-----------------------------------------------')

print(classification_report(y_test,y_pred))

Accuracy score :  0.7402597402597403
-----------------------------------------------
[[81 18]
 [22 33]]
-----------------------------------------------
              precision    recall  f1-score   support

           0       0.79      0.82      0.80        99
           1       0.65      0.60      0.62        55

    accuracy                           0.74       154
   macro avg       0.72      0.71      0.71       154
weighted avg       0.74      0.74      0.74       154



### Observations when tuning params

- With sigmoid, it has a 74% accuracy

- ReLu -> 75%

- Increased iterations to 500 for ReLu and performance dropped to 71%

- With tanh and performance increased to 77% !

* will need clarification(research)